### DISCUSSION TITLE
EMQ X integration to InfluxDB

### DISCUSSION THREAD

**Discussion by timhallinflux** [emqx/emqx#5015] [Ideas]
The current integration approach between EMQ X and InfluxDB is based on two things:
- UDP
- JSON formatted messages

InfluxDB OSS v1.x supports both of these.  However, InfluxDB Cloud does not support either.  InfluxDB OSS v2 has been released since November and also does not support UDP or JSON formatted messages.

Ideally, we could work together to upgrade the integration to leverage an HTTP-based integration between EMQ X which would allow for users to use either InfluxDB OSS v2 or InfluxDB Cloud.

**Comment by zmstone** [comment]
Hi @timhallinflux 
Thank you for reporting this.
Is http api the only way forward with new version influxdb?

**Comment by timhallinflux** [comment]
Yes, for now. We are working on a native MQTT ingestion mechanism where you would configure InfluxDB Cloud to listen to the MQTT broker.  That is planned for 2H21.  

If you wanted to leverage Telegraf (our distributed agent which is open source), you could embed those bits into EMQ X and simply use your configuration approach to pull from the MQTT broker and then configure the appropriate InfluxDB output.

Happy to explain this approach in more detail... if you wish.

**Reply by zmstone** [reply]
Hi @timhallinflux @Benniu and I are interested to learn more details about how we can integrate influxDB.
Maybe we can have a chat over a call ? You can reach me at the email from my github commit history ;)

**Comment by zmstone** [comment]
sorry for the delayed reply.
we have been testing a bit with webhook.

converting this thread to a discussion.

**Comment by zibuyu1995** [comment]
We use EMQ X Webhook to persist data to the Influxdb Cloud:

### Create Bucket
After logging in to the influxDB's console, go to `Load Data` page and create a new bucket.
![bucket](https://user-images.githubusercontent.com/17525759/122256297-80567a00-cf01-11eb-9ebd-542f700ad4dd.png)


### Generate Token

 Go back to `Load Data` page and generate a new token. At this time, we will generate a token with full access.
   
![token_done](https://user-images.githubusercontent.com/17525759/122256324-86e4f180-cf01-11eb-8462-e63d65b7bca1.png)



### Create EMQ X Webhook Resource

 Resource url: https://us-east-1-1.aws.cloud2.influxdata.com/api/v2/write
 
<img width="1674" alt="webhook" src="https://user-images.githubusercontent.com/17525759/122256383-9a905800-cf01-11eb-8e5b-b11841fbe981.png">




### Create EMQ X Rule SQL

In this example we demonstrate how to store temperature and humidity sensor data to the influxdb cloud.

   ```sql
   SELECT
     payload.location as location,
     payload.temp as temp,
     payload.hum as hum
   FROM
     "emqx/test"
   ```
   
<img width="1649" alt="rule_sql" src="https://user-images.githubusercontent.com/17525759/122256443-a8de7400-cf01-11eb-9172-2487ee77dbc9.png">




### Add Rule Action

#### Add Request Parameters

Need to change `{organizations}` and `{bucket}` 

* Method: POST
* Path: /api/v2/write?org={organizations}&bucket={bucket}&precision=ns

#### Add Request Headers

Authorization needs to be modified. Authorization is the token generated by Influxdb Cloud

* Content-Encoding: identity
* Content-Type: text/plain
* Accept: application/json
* Authorization: Token _kC8WcOxoxxjH0RAqHT-uHQD5T69RC-fvTEOX4WJMYbwNLH3pFyjQD-7JiEYAxGi84UubA9N9l0h4zkWJHB-T4w==
* User-Agent: emqx

#### Add Request Body
* Measurement: temp_hum
* Tags: location
* Fields: temp, hum
```
temp_hum,location=${location} temp=${temp},hum=${hum}
```
<img width="1679" alt="action" src="https://user-images.githubusercontent.com/17525759/122256563-c7446f80-cf01-11eb-987d-497d96d8ccef.png">




### Publish temperature and humidity data

<img width="1025" alt="mqttx" src="https://user-images.githubusercontent.com/17525759/122256587-cb708d00-cf01-11eb-8209-a5729d317c24.png">




### View EMQ X Rule Matched

<img width="1662" alt="monitor" src="https://user-images.githubusercontent.com/17525759/122256620-d2979b00-cf01-11eb-9533-ae97bdcae727.png">




### View data in Influxdb Cloud

<img width="1668" alt="influxdb" src="https://user-images.githubusercontent.com/17525759/122256692-e3481100-cf01-11eb-9e40-ec35d1ecdbde.png">

**Comment by timhallinflux** [comment]
super!  Looks like it is all working.  Perhaps the documentation needs to be updated to reflect this approach?

**Reply by zibuyu1995** [reply]
We provide video tutorials, see the https://www.youtube.com/watch?v=1S09epPWLpc&t=43s
